R/cnd_df.R

```diff
@@ -0,0 +1,321 @@
+#' @return The input `dat` without the `cnd_df` class.
+#'
+#' @export
+rm_cnd_df <- function(dat) {
+  if (is_cnd_df(dat)) {
+    class(dat) <- class(dat)[class(dat) != "cnd_df"]
+    attr(dat, "cnd") <- NULL
+    attr(dat, "cnd_sum") <- NULL
+  }
+  return(dat)
+}
+
+#' Conditioned tibble header print method
+#'
+#' @param x A conditioned tibble of class `cnd_df`.
+#' @param ... Additional arguments passed to the default print method.
+#'
+#' @importFrom pillar tbl_sum
+#' @export
+tbl_sum.cnd_df <- function(x, ...) {
+  default_header <- NextMethod()
+
+  tally <- get_cnd_df_cnd_sum(x)
+  h2 <- sprintf("%d/%d/%d", tally[1], tally[2], tally[3])
+  c(default_header, "Cond. tbl" = h2)
+}
+
+lgl_to_chr <- function(x) {
+  ifelse(is.na(x), "-", ifelse(x, "T", "F"))
+}
+
+#' Conditioned tibble pillar print method
+#'
+#' @inheritParams pillar::ctl_new_rowid_pillar
+#' @importFrom pillar ctl_new_rowid_pillar
+#' @export
+ctl_new_rowid_pillar.cnd_df <- function(controller, x, width, ...) {
+
+  out <- NextMethod()
+  n_row <- nrow(x)
+  idx <- seq_len(n_row)
+  i <- sprintf("%d", idx)
+  i_width <- nchar(as.character(i))
+  i_max_width <- max(i_width)
+  max_width <- i_max_width + 2
+  ws <- strrep(" ", max_width - i_width - 1)
+  abb_lgl <- lgl_to_chr(attr(controller, "cnd")[idx])
+
+  row_ids <- paste0(i, ws, abb_lgl)
+  width <- max(nchar(as.character(row_ids)))
+  pillar::new_pillar(
+    list(
+      title = out$title,
+      type = out$type,
+      data = pillar::pillar_component(
+        pillar::new_pillar_shaft(list(row_ids = row_ids),
+                                 width = width,
+                                 class = "pillar_rif_shaft"
+        )
+      )
+    ),
+    width = width
+  )
+}
+
+#' Evaluate conditions
+#'
+#' @description
+#' [eval_conditions()] evaluates a set of conditions in the context of a
+#' data frame and an optional environment.
+#'
+#' The utility of this function is to provide an easy way to generate a logical
+#' vector of matching records from a set of logical conditions involving
+#' variables in a data frame (`dat`) and optionally in a supplementary
+#' environment (`.env`). The set of logical conditions are provided as
+#' expressions to be evaluated in the context of `dat` and `.env`.
+#'
+#' Variables are looked up in `dat`, then in `.env`, then in the calling
+#' function's environment, followed by its parent environments.
+#'
+#' @param dat A data frame
+#' @param ... A set of logical conditions, e.g. `y & z, x | z` (`x`, `y`, `z`
+#'   would have to exist either as columns in `dat` or in the environment
+#'   `.env`). If multiple expressions are included, they are combined with the
+#'   `&` operator.
+#' @param .na Return value to be used when the conditions evaluate to `NA`.
+#' @param .env An optional environment to look for variables involved in logical
+#'   expression passed in `...`. A data frame or a list can also be passed that
+#'   will be coerced to an environment internally.
```

Comment:
If you're passing `dat2` explicitly you don't need to indicate the environment it resides in as it is being passed on. In that case it is really just a matter of renaming `.env` to `.dat2` --- note that I am adding again the dot on the parameter name because `condition_by()` accepts `...` and having a parameter name starting with dot signals that.

Reply:
`.dat2` works well. Thank you.

Reply:
Also, let's change the order of the arguments so the user doesn't have to give additional parenthesis {} while using `condition_add`

```
assign_no_ct <-(
tgt_dat = NULL,
tgt_var,
raw_dat,
raw_var,
id_vars = oak_id_vars())

assign_ct <-(
tgt_dat = NULL,
tgt_var,
raw_dat,
raw_var,
ct_spec,
ct_clst,
id_vars = oak_id_vars())
```

But the downside is that the parameters with default values are not grouped together and they are not at the end. As you have specified, it will make the piping much better. Let's go for it.

Reply:
`.env` has been renamed to `.dat2` in `condition_add()`.